proposals/0540-jsem.rst

```diff
@@ -158,37 +158,45 @@ Implementation of the protocol
-   fewer CPU cores than its capabilities: the stop-the-world cost of garbage
-   collection becomes much more expensive. If we were to give GHC a fixed number
-   ``N`` of capabilities, then parallel garbage collections in that GHC would
-   recruit ``N`` OS threads. Instead we propose that GHC will adjust its number
-   of capabilities, via ``setNumCapabilities`` [8]_ , to the number of modules
-   it is compiling.
-   Note that GHC already calls ``setNumCapabilities`` with the argument to ``-j``.
-
-   We have concerns that rapidly adjusting the number of capabilities may have
-   adverse effects: we propose that GHC will limit the rate at which it can
-   release tokens to the semaphore so as to reduce this thrashing.
-   The exact mechanism of rate limiting is unspecified here.
-
-B. We believe that GHC invocations should not be too eager to release
-   semaphore tokens. This skews the balance in favour of in-unit parallelism
-   (1 unit with many capabilities), and against compiling many units in parallel
-   (many units with a single capability).
-   The justification is that the memory used by compiling units can be released
-   before starting another parallel process. Were GHC to release semaphore tokens
-   too eagerly, it could end up compiling a large number of packages in parallel
-   which each have a large loaded EPS. Combined, this will use a significant amount
-   of memory.
-
-   To achieve this, we propose a minimum delay between acquiring a resource and
-   releasing a resource. This allows us to prioritise completing a single large
-   package before moving on to other work. The exact mechanism through which this
-   is achieved is left unspecified in this proposal.
+The GHC jobclient implementation would have the following characteristics:
+
+A. GHC requests tokens for each module it is compiling concurrently.
+   (The current implementation requests one token per module, but it could make
+   sense to request more, e.g. if one is able to parallelise the simplifier
+   workload.)
+
+B. GHC always returns all the tokens it has acquired from the semaphore,
+   either upon successful completion or when an exception is raised,
+   by using the `bracket` pattern.
+
+C. GHC should adjust its number of capabilities, via ``setNumCapabilities`` [8]_ ,
+   to the number of tokens it is using (up to the number of available CPU cores).
+   This is because there is a hidden cost in having a GHC program
+   (such as ``ghc`` itself) run on fewer CPU cores than its capabilities: the
+   stop-the-world cost of garbage collection becomes much more expensive.
+   If we were to give GHC a fixed number ``N`` of capabilities, then parallel
+   garbage collections in that GHC would recruit ``N`` OS threads.
+   Note that GHC already calls ``setNumCapabilities`` with the argument
+   to ``-j``.
+
+D. GHC should rate-limit the release of semaphore tokens (the precise mechanism
```

Comment:
Does your current implementation have a view on how/when to release semaphore tokens?

I'm asking not because I think this should be specified (I agree it's an implementation detail), but because it seems like a very subtle and tricky knob to get right. That's a cost that the GHC Maintainers will have to bear and understand, so it would help if you already have a framework for how to reason about it.